torch/cuda/graphs.py

```diff
@@ -0,0 +1,289 @@
+                                           static_grad_inputs):
+            class Graphed(torch.autograd.Function):
+                @staticmethod
+                def forward(ctx, *inputs):
+                    with torch.no_grad():
+                        for i, arg in zip(static_input_surface, inputs):
+                            if i.data_ptr() != arg.data_ptr():
+                                i.copy_(arg)
+                    fwd_graph.replay()
+                    assert isinstance(static_outputs, tuple)
+                    return tuple(o.detach() for o in static_outputs)
+
+                @staticmethod
+                def backward(ctx, *grads):
+                    with torch.no_grad():
+                        for g, grad in zip(static_incoming_grads, grads):
+                            if g is None:
+                                assert grad is None
+                            else:
+                                # don't copy if autograd gods have been kind and the
+                                # incoming grad is already in the right place
+                                if g.data_ptr() != grad.data_ptr():
+                                    g.copy_(grad)
+                    bwd_graph.replay()
+
+                    # Input args that didn't require grad expect a None gradient.
+                    return tuple(b.detach() if b is not None else b for b in static_grad_inputs)
+
+            def functionalized(*user_args):
+                out = Graphed.apply(*(user_args + module_params))
+                return out[0] if output_was_tensor else out
+
+            return functionalized
+
+        # Put together the final graphed callables
+        ret = []
+        for i, func in enumerate(callables):
+            graphed = make_graphed_autograd_function(fwd_graphs[i],
+                                                     bwd_graphs[i],
+                                                     per_callable_module_params[i],
+                                                     per_callable_output_was_tensor[i],
+                                                     per_callable_static_input_surfaces[i],
+                                                     per_callable_static_outputs[i],
+                                                     per_callable_static_incoming_grads[i],
+                                                     per_callable_static_grad_inputs[i])
+
+            if isinstance(func, torch.nn.Module):
+                def make_graphed_forward(func, training, graphed, orig_fwd):
+                    def new_fwd(*user_args):
+                        # If the module's training-or-eval state matches what we graphed,
+                        # run the graph, otherwise run the original forward method
+                        if func.training == training:
+                            return graphed(*user_args)
+                        else:
+                            return orig_fwd(*user_args)
```

Comment:
I process every callable the user passes. Why would I want to skip one?

Reply:
Considering 3 modules, a regular evaluation would be:
- eval mod1 graph
- eval mod2 graph
- eval mod3 graph

But if you change the training mode of mod2, you end up doing
- eval mod1 graph
- eval mod2 real forward
- eval mod3 graph  <- is that one ok? The graph for mod2 didn't run before this one!

Reply:
Yes, that should be ok. The fact that "eval mod2 real forward" uses memory from the main eager pools instead of the graph pools should not affect "eval mod3 graph".

Of course, the user is still responsible for passing "eval mod2 real forward"'s output to "eval mod3 graph", but that's also not a problem. "eval mod3 graph" always copies new inputs to its graph's static input memory as needed.

Reply:
Sounds good!
I just wanted to confirm that the shared memory pool won't be affected by that!